### PR TITLE
better error messages for substr hooks

### PR DESCRIPTION
**Pull request by theo25**
This PR improves the error messages for the `substrString` hook in order to address #493

### CODE REVIEW

**Comment by theo25** [comment]
I wanted to have something that simulates the `std::format` of C++20. I guess I can make the code simpler, but I don't see how we can print detailed error messages without some kind of string formatting

**Comment by theo25** [comment]
Would it help, if I add some unit tests that test the string formatting code?

File: runtime/strings/bytes.cpp [hunk 1]
```diff
@@ -8,6 +9,18 @@
...
 #include "runtime/header.h"
 
+template <typename... Args>
+std::string string_format(const std::string &format, Args... args) {
+  int size_s = std::snprintf(nullptr, 0, format.c_str(), args...) + 1;
+  if (size_s <= 0) {
+    throw std::runtime_error("Error during formatting.");
+  }
+  size_t size = static_cast<size_t>(size_s);
+  std::unique_ptr<char[]> buf(new char[size]);
```
Comment (Baltoli) [inline runtime/strings/bytes.cpp:19]: Very minor point but I tend to prefer `make_unique` here

**Review by Baltoli — APPROVED** [review]
One small point but otherwise looks good!

**Comment by theo25** [comment]
I have changed the implementation to use `std::string` and string concatenation to generate the error messages in a less complicated manner.

File: runtime/strings/bytes.cpp [hunk 2]
```diff
@@ -122,10 +122,22 @@ SortBytes hook_BYTES_substr(SortBytes input, SortInt start, SortInt end) {
   uint64_t ustart = get_ui(start);
   uint64_t uend = get_ui(end);
   if (uend < ustart) {
-    throw std::invalid_argument("Invalid string slice");
+    throw std::invalid_argument(
+        std::string("Invalid string slice for string \"")
+        + std::string(input->data) + std::string("\": Requested start index ")
```
Comment (Baltoli) [inline runtime/strings/bytes.cpp:127]: I know we're trying to keep the code simple, but what if the string data is huge? Won't this produce a potentially enormous error message?

**Review by Baltoli — CHANGES_REQUESTED** [review]

File: runtime/strings/bytes.cpp (hunk 2 above)
Comment (theo25) [reply runtime/strings/bytes.cpp]: That's fair, I removed the string from the error message. It should still be more informative than before.

**Review by Baltoli — APPROVED** [review]

**Pull request merged**